flink-autoscaler/src/main/java/org/apache/flink/autoscaler/utils/MemoryTuningUtils.java

```diff
@@ -0,0 +1,132 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.autoscaler.utils;
+
+import org.apache.flink.autoscaler.JobAutoScalerContext;
+import org.apache.flink.autoscaler.ScalingSummary;
+import org.apache.flink.autoscaler.config.AutoScalerOptions;
+import org.apache.flink.autoscaler.metrics.EvaluatedMetrics;
+import org.apache.flink.autoscaler.metrics.EvaluatedScalingMetric;
+import org.apache.flink.autoscaler.metrics.ScalingMetric;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+
+/** Tunes the TaskManager memory. */
+public class MemoryTuningUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MemoryTuningUtils.class);
+
+    public static Optional<MemorySize> tuneTaskManagerHeapMemory(
+            JobAutoScalerContext<?> context,
+            EvaluatedMetrics evaluatedMetrics,
+            Map<JobVertexID, ScalingSummary> scalingSummaries) {
+
+        var config = context.getConfiguration();
+        if (!config.get(AutoScalerOptions.MEMORY_TUNING_ENABLED)) {
+            return Optional.empty();
+        }
+
+        var globalMetrics = evaluatedMetrics.getGlobalMetrics();
+        double avgHeapSize = globalMetrics.get(ScalingMetric.HEAP_AVERAGE_SIZE).getAverage();
+
+        double numTaskSlotsUsed = globalMetrics.get(ScalingMetric.NUM_TASK_SLOTS_USED).getCurrent();
+        int taskSlotsPerTm = config.get(TaskManagerOptions.NUM_TASK_SLOTS);
+        int currentNumTMs = (int) Math.ceil(numTaskSlotsUsed / taskSlotsPerTm);
+
+        double usedTotalHeapSize = currentNumTMs * avgHeapSize;
+        LOG.info("Total used heap size: {}", new MemorySize((long) usedTotalHeapSize));
+        usedTotalHeapSize *= computeDataChangeRate(evaluatedMetrics);
+        LOG.info("Resized total heap size: {}", new MemorySize((long) usedTotalHeapSize));
+
+        int numTaskSlotsAfterRescale =
+                ResourceCheckUtils.estimateNumTaskSlotsAfterRescale(
+                        evaluatedMetrics, scalingSummaries, numTaskSlotsUsed);
+        int newNumTms = (int) Math.ceil(numTaskSlotsAfterRescale / (double) taskSlotsPerTm);
+        LOG.info(
+                "Estimating {} task slots in use after rescale, spread across {} TaskManagers",
+                numTaskSlotsAfterRescale,
+                newNumTms);
+
+        MemorySize newHeapSize = new MemorySize((long) (usedTotalHeapSize / newNumTms));
+        // TM container memory can never grow beyond the user-specified max
+        Optional<MemorySize> maxMemory = context.getTaskManagerMemoryFromSpec();
```

Comment:
> From the code, I understand that "Adjust memory" will only adjust the memory down to below the user-specified memory, but not above the user-specified memory, right?

That is correct.

> I'm not sure could we make this option name as generic as possible? Such as: `job.autoscaler.taskmanager-memory.scaling.enabled` and `job.autoscaler.taskmanager-memory.scaling.heap.min`.

Sure, we can change those. Maybe `job.autoscaler.memory.taskmanager.heap.min` would be even more generic? It would also allow to add a jobmanager section under memory.

> In the short term, we only turn down. In the long term, we might support turn up the memory. If the option name is more generic, it's easy to compatible. WDYT?

I think the current option prefix would also allow to support increasing in the future, e.g. by adding `*.heap.max` option.